bliss/backbone.py

```diff
@@ -0,0 +1,135 @@
+# flake8: noqa # pylint: disable-all
+import contextlib
+import math
+from copy import deepcopy
+from pathlib import Path
+
+import torch
+import yaml
+from torch import nn  # needed to parse config
+from yolov5.models.yolo import C3, SPPF, BaseModel, Bottleneck, Concat, Conv, Detect
+from yolov5.utils.autoanchor import check_anchor_order
+from yolov5.utils.general import make_divisible
+
+# needed to parse config
+from bliss.modules import Conv3d, Conv3d_down
+
+
+class Backbone(BaseModel):
+    def __init__(self, cfg="yolov5s.yaml", ch=3, nc=None, anchors=None, n_imgs=2):
+        super().__init__()
+        if isinstance(cfg, dict):
+            self.yaml = cfg  # model dict
+        else:  # is *.yaml
+            self.yaml_file = Path(cfg).name
+            with open(cfg, encoding="ascii", errors="ignore") as f:
+                self.yaml = yaml.safe_load(f)  # model dict
+
+        # Define model
+        # input channels
+        ch = self.yaml["ch"] = self.yaml.get("ch", ch)
+        self.num_imgs = n_imgs
+        self.model, self.save = parse_model(
+            deepcopy(self.yaml), ch=[ch], n_imgs=self.num_imgs
+        )  # model, savelist
+        self.names = [str(i) for i in range(self.yaml["nc"])]  # default names
+        self.inplace = self.yaml.get("inplace", True)
+        # Build strides, anchors
+        m = self.model[-1]  # Detect()
+        if isinstance(m, (Detect)):
+            s = 256  # 2x min stride
+            m.inplace = self.inplace
+
+            # NOTE: 5-D input
+            d = self.num_imgs  # set d = 2 in 5-band case
+            m.stride = torch.tensor(
+                [s / x.shape[-2] for x in self.forward(torch.zeros(1, ch, d, s, s))]
+            )
+
+            check_anchor_order(m)
+            m.anchors /= m.stride.view(-1, 1, 1)
+            self.stride = m.stride
+            self._initialize_biases()  # only run once
+
+    def forward(self, x, profile=False, visualize=False):
+        return self._forward_once(x, profile, visualize)  # single-scale inference, train
+
+    def _initialize_biases(self, cf=None):  # initialize biases into Detect(), cf is class frequency
+        # https://arxiv.org/abs/1708.02002 section 3.3
+        m = self.model[-1]  # Detect() module
+        for mi, s in zip(m.m, m.stride):  # from
+            b = mi.bias.view(m.na, -1)  # conv.bias(255) to (3,85)
+            b.data[:, 4] += math.log(8 / (640 / s) ** 2)  # obj (8 objects per 640 image)
+            b.data[:, 5 : 5 + m.nc] += (
+                math.log(0.6 / (m.nc - 0.99999)) if cf is None else torch.log(cf / cf.sum())
+            )  # cls
+            mi.bias = torch.nn.Parameter(b.view(-1), requires_grad=True)
+
+
+def parse_model(d, ch, n_imgs):  # model_dict, input_channels(3)
```

Comment:
Please add a comment explaining that this isn't our code in this file: it's a modified version of code from yolov5. (We want to give them credit.)

Reply:
It also would be helpful to future developers to explain why we're including yolov5 code here

Reply:
Will do!